### PR TITLE
cluster: storage node addesses discovery, added vmauth

### PR DESCRIPTION
**Pull request by AndrewChubatiuk**
- added ability to set additional storage nodes to default ones using `<vmselect|vminsert>.extraArgs.storageNode` array
- added storage node srv discovery if enterprise license is enabled. fixes https://github.com/VictoriaMetrics/helm-charts/issues/273
- added optional vmauth in front of vmselect and vminsert. fixes https://github.com/VictoriaMetrics/helm-charts/issues/376

related issue https://github.com/VictoriaMetrics/VictoriaMetrics/issues/7040

### CODE REVIEW

**Review by Haleygo — COMMENTED** [review]
>added optional vmauth in front of vmselect and vminsert. fixes https://github.com/VictoriaMetrics/helm-charts/issues/376

User can only benefit more from vmauth than k8s service when address under `url_prefix` is pointing to a headless service, that way vmauth can resolve all the backend addresses in advance and proxy request to the least-loaded endpoint. But headless svc is not the default service type for both vmselect and vminsert here.
And I don't think we should add more logic to change the service type when enable the vmauth without making user understand why we do that. What about adding some instructions in readme?

File: charts/victoria-metrics-cluster/values.yaml [hunk 1]
```diff
@@ -11,12 +11,18 @@ global:
...
+  # -- k8s cluster domain suffix, uses for building storage pods' FQDN. Details are [here](https://kubernetes.io/docs/tasks/administer-cluster/dns-custom-nameservers/)
+  cluster:
+    dnsDomain: cluster.local
 
 # -- K8s cluster domain suffix, uses for building storage pods' FQDN. Details are [here](https://kubernetes.io/docs/tasks/administer-cluster/dns-custom-nameservers/)
 clusterDomainSuffix: cluster.local
 # -- Print information after deployment
 printNotes: true
 
+# use SRV discovery for storageNode and selectNode flags for enterprise version
```
[the remark this answers is not in the document]
Comment (AndrewChubatiuk) [reply charts/victoria-metrics-cluster/values.yaml:23]: license can also be defined at `global.license`

[the remark this answers is not in the document]
Comment (AndrewChubatiuk) [reply charts/victoria-metrics-cluster/values.yaml:23]: added failure if no license is defined with autoDiscovery enabled

**Comment by AndrewChubatiuk** [comment]
added for `vmauth.enabled` param description about headless services

**Comment by Haleygo** [comment]
Need to set `vmselect.name` to "" for the new name template to avoid creating new service.
https://github.com/VictoriaMetrics/helm-charts/blob/d478a2454d81cea5025e16f5b1c02289751a7326/charts/victoria-metrics-cluster/values.yaml#L71

btw, templates seem to complicate debugging and make value modifications more fragile. 
I don't understand why we need to consolidate all the component' services into one template yaml, and they have very different port settings in the `_helpers.tpl`. The previous resources look easier to me only with some duplicated fields.
https://github.com/VictoriaMetrics/helm-charts/blob/d478a2454d81cea5025e16f5b1c02289751a7326/charts/victoria-metrics-cluster/templates/_helpers.tpl#L261-L288

File: charts/victoria-metrics-cluster/templates/_helpers.tpl [hunk 2]
```diff
@@ -106,131 +67,135 @@ app: {{ .Values.vmselect.name }}
...
+  {{- $_ := set . "style" "plain" }}
+  {{- $_ := set . "appKey" "vmstorage" }}
+  {{- $args = mergeOverwrite $args (fromYaml (include "vm.license.flag" .)) -}}
+  {{- $args = mergeOverwrite $args $app.extraArgs -}}
+  {{- $storage := $Values.vmstorage }}
+  {{- if and (not $app.suppressStorageFQDNsRender) (and $storage.enabled $storage.replicaCount) }}
+    {{- $storageNodes := default list }}
+    {{- $fqdn := include "vm.fqdn" . }}
+    {{- if $Values.autoDiscovery }}
+      {{- if (include "vm.enterprise.disabled" . ) "true" }}
```
Comment (zekker6) [inline charts/victoria-metrics-cluster/templates/_helpers.tpl:104]:
```suggestion
      {{- if eq (include "vm.enterprise.disabled" . ) "true" }}
```

File: charts/victoria-metrics-cluster/templates/_helpers.tpl [hunk 3]
```diff
@@ -106,131 +67,135 @@ app: {{ .Values.vmselect.name }}
...
+  {{- $_ := set . "appKey" "vmstorage" }}
   {{- $_ := set $args "cacheDataPath" $app.cacheMountPath -}}
   {{- $args = mergeOverwrite $args (fromYaml (include "vm.license.flag" .)) -}}
   {{- $args = mergeOverwrite $args $app.extraArgs -}}
+  {{- $storage := $Values.vmstorage }}
+  {{- if and (not $app.suppressStorageFQDNsRender) (and $storage.enabled $storage.replicaCount) }}
+    {{- $storageNodes := default list }}
+    {{- $fqdn := include "vm.fqdn" . }}
+    {{- if $Values.autoDiscovery }}
+      {{- if (include "vm.enterprise.disabled" . ) "true" }}
```
Comment (zekker6) [inline charts/victoria-metrics-cluster/templates/_helpers.tpl:150]:
```suggestion
      {{- if eq (include "vm.enterprise.disabled" . ) "true" }}
```

File: charts/victoria-metrics-cluster/templates/_helpers.tpl [hunk 4]
```diff
@@ -106,131 +67,135 @@ app: {{ .Values.vmselect.name }}
...
+      {{- $_ := unset . "appIdx" }}
+    {{- end }}
+    {{- $_ := set $args "storageNode" (concat ($args.storageNode | default list) $storageNodes) }}
+  {{- end }}
+  {{- if and $app.statefulSet.enabled $app.enabled $app.replicaCount }}
+    {{- $selectNodes := default list }}
+    {{- $_ := set . "appKey" "vmselect" }}
+    {{- $fqdn := include "vm.fqdn" . }}
+    {{- if $Values.autoDiscovery }}
+      {{- if (include "vm.enterprise.disabled" . ) "true" }}
```
Comment (zekker6) [inline charts/victoria-metrics-cluster/templates/_helpers.tpl:171]:
```suggestion
      {{- if eq (include "vm.enterprise.disabled" . ) "true" }}
```

**Review by zekker6 — COMMENTED** [review]
LGTM in general.
Added a few suggestions to fix a typo in configs. I would expect CI to detect this, but it seems like we don't have a test set to cover this templating path. Could you add some tests to cover this part?

**Comment by AndrewChubatiuk** [comment]
> Could you add some tests to cover this part?
thanks for review. added this scenario

**Review by zekker6 — APPROVED** [review]

**Pull request merged**